### PR TITLE
Integrate with Kubernetes API

### PR DESCRIPTION
**Pull request by stuartleeks**
Initial WIP for drilling into the Kubernetes API for AKS clusters

TODO list:
 - [X] get cluster credentials
 - [X] get cluster swagger
 - [X] create nodes from swagger dynamically
 - [x] reuse swagger parsing when revisiting root k8s api node
 - [ ] testing!
 - [x] Remove duplicate nodes under `/api/v1`
 - [x] handle deletion
 - [X] refactoring/renaming classes now that they are better understood
 - [X] colorise YAML
 - [ ] Handle updates for non ARM resources
 - [x] ~~stretch/future PR:~~ handle logs requiring extra parameter for container name
 - [~~stretch/future PR: add command to copy node IDs (e.g. for use in `navigate` command) UPDATE - this is #21~~

### CODE REVIEW

**Comment by stuartleeks** [comment]
I can't decide whether to add code to avoid requerying the swagger etc for a cluster. Perf doesn't seem too bad, and requerying means that you would pick up any new CRDs which could be a benefit. Any thoughts @lawrencegripper?

**Comment by stuartleeks** [comment]
Cool, I'll try to take a look at the cache stuff.

Let me know how your general testing goes

**Comment by stuartleeks** [comment]
OK- latest commit has caching of the k8s swagger etc

**Comment by lawrencegripper** [comment]
WOW - I want this merged so badly!

image

image

File: internal/pkg/handlers/containerService-swaggerApiSet.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+		data := string(buf)
+		return data, nil
+	}
+	return "", fmt.Errorf("Response failed with %s (%s)", response.Status, url)
+}
+
+// ExpandResource returns metadata about child resources of the specified resource node
+func (c SwaggerAPISetContainerService) ExpandResource(ctx context.Context, currentItem *TreeNode, resourceType swagger.ResourceType) (APISetExpandResponse, error) {
+
+	if resourceType.Endpoint.TemplateURL == "/api/v1/namespaces/{namespace}/pods/{name}/log" {
```
Comment (lawrencegripper) [inline internal/pkg/handlers/containerService-swaggerApiSet.go]: Very cool that you can get logs :)

File: internal/pkg/handlers/containerService.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+	return false, nil
+}
+
+// Expand returns ManagementPolicies in the StorageAccount
+func (e *AzureKubernetesServiceExpander) Expand(ctx context.Context, currentItem *TreeNode) ExpanderResult {
+
+	swaggerResourceType := currentItem.SwaggerResourceType
+	if currentItem.Namespace != "AzureKubernetesService" &&
+		swaggerResourceType != nil &&
+		swaggerResourceType.Endpoint.TemplateURL == "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ContainerService/managedClusters/{resourceName}" {
```
Comment (lawrencegripper) [inline internal/pkg/handlers/containerService.go]: nit: maybe pull out into a const

Comment (lawrencegripper) [reply internal/pkg/handlers/containerService.go]: Changed my mind... think it'd be less clear as a const

Comment (stuartleeks) [reply internal/pkg/handlers/containerService.go]: #done

**Comment by lawrencegripper** [comment]
@stuartleeks you happy for this to merge?

**Comment by stuartleeks** [comment]
yup, happy to merge - have some more improvements on it but happy to put them in a new PR

**Comment by stuartleeks** [comment]
Need a review btw ;-)

File: internal/pkg/handlers/containerService.go [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+		},
+		SuppressAPIVersion: true,
+	}
+	var paths []*swagger.Path
+	paths, err = swagger.MergeSwaggerDoc(paths, &config, doc, false)
+	if err != nil {
+		return swaggerResourceTypes, err
+	}
+
+	// uncomment this code to help debug the generated API paths!
```
Comment (jjcollinge) [inline internal/pkg/handlers/containerService.go]: @lawrencegripper what's the wider story about enabling debug features - do we drive all this behaviour via the `--debug` flag? Do we need to look for build macro alternatives or do you think uncomment and rebuild is sufficient?

File: internal/pkg/handlers/containerService.go (hunk 2 above)
Comment (stuartleeks) [reply internal/pkg/handlers/containerService.go]: It is repeated in a couple of places, so I opted for const :-)

File: internal/pkg/handlers/containerService.go (hunk 3 above)
Comment (stuartleeks) [reply internal/pkg/handlers/containerService.go]: Just pushed an update that outputs this based on whether debug is set

**Comment by stuartleeks** [comment]
YAML colourisation just snuck in :-)

**Comment by lawrencegripper** [comment]
I'm declaring this PR a MONSTER!

**Pull request merged**